### PR TITLE
create launcher and worker togther

### PR DESCRIPTION
**Pull request by davidstack**
for issue #131 ,this commit will create lancher and worker togther, in the init container of lancher, the init container will wait the worker untill running

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/mpi-operator/132)

### CODE REVIEW

File: cmd/kubectl-delivery/deliver_kubectl.sh [hunk 1]
```diff
@@ -22,3 +22,42 @@ fi
 : ${TARGET_DIR:?"Need to set TARGET_DIR, e.g. /opt/kube"}
 
 cp $(which kubectl) ${TARGET_DIR}
+
```
Comment (gaocegege) [inline cmd/kubectl-delivery/deliver_kubectl.sh]: What's the purpose of the script here?

Comment (davidstack) [reply cmd/kubectl-delivery/deliver_kubectl.sh]: to check the worker  is running, if the worker is in other state, the init contianer will be in loop,just waiting for the worker running

**Comment by davidstack** [comment]
@gaocegege  in my dev env, i test ok ,it will create lancher and worker togther (waiting kube-batch schedule ).
the init continer log is 
`[root@node3 ~]# docker logs -f e953b1fef837

caffe-mpi-worker-0 slots=2
1
caffe-mpi-worker-1 slots=2
2
worker name is caffe-mpi-worker-0
Unable to use a TTY - input is not a terminal or the right kind of file
hello is hello
caffe-mpi-worker-0 is running..
worker name is caffe-mpi-worker-1
Unable to use a TTY - input is not a terminal or the right kind of file
hello is hello
caffe-mpi-worker-1 is running..`

the status of mpi woker and lancher is

`[root@node2 mpi]# kubectl get pods -o wide

NAME                       READY   STATUS              RESTARTS   AGE   IP              NODE    NOMINATED NODE   READINESS GATES
caffe-mpi-launcher-pssbh   0/1     Init:0/1            0          2s    <none>          node2   <none>           <none>
caffe-mpi-worker-0         0/1     ContainerCreating   0          2s    <none>          node2   <none>           <none>
caffe-mpi-worker-1         0/1     ContainerCreating   0          2s    <none>          node3   <none>           <none>
`

**Comment by gaocegege** [comment]
@davidstack 

Thanks, I understand the meaning of the script.

BTW, @terrytangyuan @wackxu Please take a look.

**Comment by wackxu** [comment]
@davidstack The CI is failed, Could you fix the CI first?

File: pkg/controllers/v1alpha2/mpi_job_controller.go [hunk 2]
```diff
@@ -530,15 +531,19 @@ func (c *MPIJobController) syncHandler(key string) error {
...
-		if workerReady && launcher == nil {
-			launcher, err = c.kubeClient.BatchV1().Jobs(namespace).Create(c.newLauncher(mpiJob, c.kubectlDeliveryImage))
-			if err != nil {
+	
+		launcher, err = c.kubeClient.BatchV1().Jobs(namespace).Create(c.newLauncher(mpiJob, c.kubectlDeliveryImage))
+		if err != nil {
 				return err
 			}
-		}
+		// If the worker is ready, start the launcher.
```
Comment (terrytangyuan) [inline pkg/controllers/v1alpha2/mpi_job_controller.go]: Remove the commented code since it's no longer needed?

Comment (davidstack) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: ok

**Comment by davidstack** [comment]
@wackxu  i find three test function failed. 
TestLauncherRestarting
TestLauncherDoesNotExist
TestLauncherActive

but it always has this error
error syncing mpi job: jobs.batch "test-launcher" already exists

could you give me some suggestion to change this ci test? thanks

**Comment by davidstack** [comment]
@wackxu  i change the CI ，and all tests PASSED，but it still have this error

`The command "gometalinter --config=linter_config.json --vendor ./..." exited with 1.`
any suggestion？ thanks

**Comment by wackxu** [comment]
@davidstack You can try to run ` gofmt -s -w yourfile` to fix this.

**Comment by davidstack** [comment]
@wackxu thanks, everything is ok.

File: pkg/controllers/v1alpha2/mpi_job_controller.go [hunk 3]
```diff
@@ -530,15 +530,14 @@ func (c *MPIJobController) syncHandler(key string) error {
 		if err != nil {
 			return err
 		}
-
-		// If the worker is ready, start the launcher.
-		workerReady := worker.Status.ReadyReplicas == workerReplicas
-		if workerReady && launcher == nil {
+		_, lancherErr := c.kubeClient.BatchV1().Jobs(namespace).Get(mpiJob.Name+launcherSuffix, metav1.GetOptions{})
```
Comment (wackxu) [inline pkg/controllers/v1alpha2/mpi_job_controller.go]: We do not need get mpijob from kube-apiserver again, we have just get launcher above from the lister, so just judge whether the launcher is nil.

File: pkg/controllers/v1alpha2/mpi_job_controller.go [hunk 4]
```diff
@@ -530,15 +530,14 @@ func (c *MPIJobController) syncHandler(key string) error {
...
+		if errors.IsNotFound(lancherErr) {
 			launcher, err = c.kubeClient.BatchV1().Jobs(namespace).Create(c.newLauncher(mpiJob, c.kubectlDeliveryImage))
 			if err != nil {
 				return err
 			}
 		}
+
```
Comment (wackxu) [inline pkg/controllers/v1alpha2/mpi_job_controller.go]: Remove this blank line

File: pkg/controllers/v1alpha2/mpi_job_controller_test.go [hunk 5]
```diff
@@ -415,6 +415,9 @@ func (f *fixture) expectCreateJobAction(d *batchv1.Job) {
 func (f *fixture) expectUpdateJobAction(d *batchv1.Job) {
 	f.kubeActions = append(f.kubeActions, core.NewUpdateAction(schema.GroupVersionResource{Resource: "jobs"}, d.Namespace, d))
 }
+func (f *fixture) expectGetJobAction(d *kubeflow.MPIJob) {
```
Comment (wackxu) [inline pkg/controllers/v1alpha2/mpi_job_controller_test.go]: Add one blank line before the func

File: pkg/controllers/v1alpha2/mpi_job_controller.go (hunk 3 above)
Comment (davidstack) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: got it

File: pkg/controllers/v1alpha2/mpi_job_controller.go (hunk 4 above)
Comment (davidstack) [reply pkg/controllers/v1alpha2/mpi_job_controller.go]: got it

File: pkg/controllers/v1alpha2/mpi_job_controller_test.go (hunk 5 above)
Comment (davidstack) [reply pkg/controllers/v1alpha2/mpi_job_controller_test.go]: got it

**Comment by davidstack** [comment]
@wackxu  have any other problem？

**Comment by wackxu** [comment]
@davidstack Could you rebase the commit to one?  then lgtm

**Comment by davidstack** [comment]
@wackxu the commit is one commit now,thanks

File: cmd/kubectl-delivery/deliver_kubectl.sh [hunk 6]
```diff
@@ -22,3 +22,42 @@ fi
...
+        if [[ "a${status}" == "a${hello}" ]] ;then
+            echo "${worker} is running.."
+            break
+        fi
+        sleep 1
+     done
+done;
+
+#/etc/mpi/hostfile
+#caffe-mpi-worker-0 slots=1
```
Comment (terrytangyuan) [inline cmd/kubectl-delivery/deliver_kubectl.sh]: Is this needed? Please remove other unneeded code/comment as well.

**Pull request merged**